File: attocode/src/shared/shared-context-state.ts
/**
 * Shared Context State
 *
 * Central failure tracker + reference pool shared across all swarm workers.
 * Workers plug into this via ContextEngineeringManager.setSharedState() so
 * they can learn from sibling failures and share compaction references.
 *
 * Reuses FailureTracker from failure-evidence.ts and Reference type from
 * reversible-compaction.ts — no new tracking logic, just shared state.
 */

import {
  FailureTracker,
  createFailureTracker,
  extractInsights,
  type FailureInput,
  type Failure,
} from '../tricks/failure-evidence.js';

import type { Reference } from '../tricks/reversible-compaction.js';

// =============================================================================
// TYPES
// =============================================================================

export interface SharedContextConfig {
  /** Maximum failures to track across all workers (default: 100) */
  maxFailures?: number;
  /** Maximum compaction references to keep (default: 200) */
  maxReferences?: number;
  /** Shared static prefix for KV-cache alignment */
  staticPrefix?: string;
}

// =============================================================================
// SHARED CONTEXT STATE
// =============================================================================

export class SharedContextState {
  private failureTracker: FailureTracker;
  private references: Map<string, Reference> = new Map();
  private maxReferences: number;
  private staticPrefix: string;

  constructor(config: SharedContextConfig = {}) {
    this.maxReferences = config.maxReferences ?? 200;
    this.staticPrefix = config.staticPrefix ?? '';

    this.failureTracker = createFailureTracker({
      maxFailures: config.maxFailures ?? 100,
      preserveStackTraces: true,
      detectRepeats: true,
      repeatWarningThreshold: 3,
    });
  }

  // ---------------------------------------------------------------------------
  // Failure tracking — workers call these, all workers see results
  // ---------------------------------------------------------------------------

  /**
   * Record a failure from a worker. The workerId is prefixed to the action
   * so cross-worker patterns become visible.
   */
  recordFailure(workerId: string, input: FailureInput): Failure {
    return this.failureTracker.recordFailure({
      ...input,
      action: `[${workerId}] ${input.action}`,
    });
  }

  /**
   * Get failure context formatted for LLM inclusion.
   */
  getFailureContext(maxFailures: number = 10): string {
    return this.failureTracker.getFailureContext({ maxFailures });
  }

  /**
   * Get actionable insights from cross-worker failures.
   */
  getFailureInsights(): string[] {
    return extractInsights(this.failureTracker.getUnresolvedFailures());
  }

  /**
   * Check if an action has failed recently (from any worker).
   */
  hasRecentFailure(action: string, withinMs: number = 60000): boolean {
    return this.failureTracker.hasRecentFailure(action, withinMs);
  }

  /**
   * Mark a failure as resolved.
   */
  resolveFailure(failureId: string): boolean {
    return this.failureTracker.resolveFailure(failureId);
  }

  /**
   * Get the underlying FailureTracker so per-worker ContextEngineeringManagers
   * can swap their local tracker for this shared one.
   */
  getFailureTracker(): FailureTracker {
    return this.failureTracker;
  }

  // ---------------------------------------------------------------------------
  // Reference pool — compaction references from all workers
  // ---------------------------------------------------------------------------

  /**
   * Add references from a worker's compaction pass.
   * Deduplicates by type:value key.
   */
  addReferences(refs: Reference[]): void {
    for (const ref of refs) {
      const key = `${ref.type}:${ref.value}`;
      if (!this.references.has(key)) {
        this.references.set(key, ref);
      }
    }

    // Enforce max references — evict oldest (first inserted)
    if (this.references.size > this.maxReferences) {
      const excess = this.references.size - this.maxReferences;
      const keys = this.references.keys();
      for (let i = 0; i < excess; i++) {
        const next = keys.next();
        if (!next.done) {
          this.references.delete(next.value);
        }
      }
    }
  }

  /**
   * Search references by query (case-insensitive substring match).
   */
  searchReferences(query: string): Reference[] {
    const lowerQuery = query.toLowerCase();
    return Array.from(this.references.values()).filter(
      (r) => r.value.toLowerCase().includes(lowerQuery),
    );
  }

  /**
   * Get all references.
   */
  getAllReferences(): Reference[] {
    return Array.from(this.references.values());
  }

  // ---------------------------------------------------------------------------
  // Cache prefix — frozen at construction
  // ---------------------------------------------------------------------------

  /**
   * Get the shared static prefix for KV-cache alignment.
   * All workers using the same prefix benefit from cache hits.
   */
  getStaticPrefix(): string {
    return this.staticPrefix;
  }

  // ---------------------------------------------------------------------------
  // Stats & cleanup
  // ---------------------------------------------------------------------------

  getStats(): { failures: number; references: number } {
    return {
      failures: this.failureTracker.getStats().total,
      references: this.references.size,
    };
  }

  clear(): void {
    this.failureTracker.clear();
    this.references.clear();
  }
}

// =============================================================================
// FACTORY
// =============================================================================

export function createSharedContextState(
  config?: SharedContextConfig,
): SharedContextState {
  return new SharedContextState(config);
}

File: src/attocode/shared/shared_context_state.py
"""Shared context state for cross-worker coordination.

Central failure tracker and reference pool shared across swarm workers.
Tracks failures, pools compaction references, and provides shared KV-cache prefix.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any

from attocode.tricks.failure_evidence import Failure, FailureTracker
from attocode.tricks.reversible_compaction import Reference


@dataclass(slots=True)
class SharedContextConfig:
    """Configuration for shared context state."""

    max_references: int = 200
    max_failures_per_worker: int = 50
    kv_cache_prefix: str | None = None


@dataclass
class SharedContextState:
    """Cross-worker failure tracking and reference pool.

    Thread-safe shared state enabling swarm workers to:
    - Report and access failure history from all workers
    - Pool compaction references for knowledge sharing
    - Use a shared KV-cache prefix for alignment
    """

    config: SharedContextConfig = field(default_factory=SharedContextConfig)

    _failure_tracker: FailureTracker = field(default_factory=FailureTracker)
    _references: list[Reference] = field(default_factory=list)
    _reference_ids: set[str] = field(default_factory=set)
    _kv_cache_prefix: str = field(default="")
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def record_failure(self, worker_id: str, failure_input: str, error: str = "") -> None:
        """Track a failure from any worker."""
        with self._lock:
            self._failure_tracker.record(
                tool_name=f"[{worker_id}]",
                input_str=failure_input,
                error=error,
            )

    def get_failure_context(self, max_failures: int = 10) -> str:
        """Format failures for LLM inclusion."""
        with self._lock:
            return self._failure_tracker.format_context(max_recent=max_failures)

    def get_failure_insights(self) -> list[str]:
        """Extract actionable insights from failures."""
        with self._lock:
            return self._failure_tracker.get_insights()

    def add_references(self, refs: list[Reference]) -> None:
        """Pool compaction references with deduplication."""
        with self._lock:
            for ref in refs:
                ref_id = f"{ref.source}:{ref.key}" if hasattr(ref, "key") else ref.source
                if ref_id not in self._reference_ids:
                    self._reference_ids.add(ref_id)
                    self._references.append(ref)

            # Trim to max
            if len(self._references) > self.config.max_references:
                removed = self._references[: len(self._references) - self.config.max_references]
                self._references = self._references[-self.config.max_references :]
                for ref in removed:
                    ref_id = f"{ref.source}:{ref.key}" if hasattr(ref, "key") else ref.source
                    self._reference_ids.discard(ref_id)

    def search_references(self, query: str) -> list[Reference]:
        """Find references by substring match."""
        with self._lock:
            query_lower = query.lower()
            return [
                r for r in self._references
                if query_lower in r.source.lower()
                or (hasattr(r, "summary") and query_lower in (r.summary or "").lower())
            ]

    @property
    def kv_cache_prefix(self) -> str:
        """Get the shared KV-cache prefix."""
        return self._kv_cache_prefix

    @kv_cache_prefix.setter
    def kv_cache_prefix(self, value: str) -> None:
        self._kv_cache_prefix = value

    def get_stats(self) -> dict[str, int]:
        """Get aggregated stats."""
        with self._lock:
            return {
                "failure_count": len(self._failure_tracker.failures),
                "reference_count": len(self._references),
            }

    def to_json(self) -> dict[str, Any]:
        """Serialize for checkpoint persistence."""
        with self._lock:
            return {
                "failures": [
                    {"tool": f.tool_name, "input": f.input_str, "error": f.error}
                    for f in self._failure_tracker.failures
                ],
                "references": [
                    {"source": r.source, "summary": getattr(r, "summary", "")}
                    for r in self._references
                ],
                "kv_cache_prefix": self._kv_cache_prefix,
            }

    def restore_from(self, data: dict[str, Any]) -> None:
        """Restore from checkpoint data."""
        with self._lock:
            self._failure_tracker = FailureTracker()
            for f in data.get("failures", []):
                self._failure_tracker.record(
                    tool_name=f.get("tool", ""),
                    input_str=f.get("input", ""),
                    error=f.get("error", ""),
                )
            self._references = []
            self._reference_ids = set()
            for r in data.get("references", []):
                ref = Reference(source=r.get("source", ""), summary=r.get("summary", ""))
                self._references.append(ref)
                self._reference_ids.add(ref.source)
            self._kv_cache_prefix = data.get("kv_cache_prefix", "")

    def clear(self) -> None:
        """Clear all state."""
        with self._lock:
            self._failure_tracker = FailureTracker()
            self._references.clear()
            self._reference_ids.clear()
            self._kv_cache_prefix = ""
